### ISSUE TITLE
V4 Planned Changes

### ISSUE THREAD

**Issue by bmish** [not-an-aardvark/eslint-plugin-eslint-plugin#120]
Apologies if these plans are already tracked somewhere, but I wanted to put together a list of the changes I'm hoping to see in the next major version. What do you think?

* Upgrade to [eslint-utils v3](https://github.com/mysticatea/eslint-utils/releases/tag/v3.0.0)
* New Node engine requirement of `"node": "12.* || 14.* || >= 16.*"` (Node 10 end-of-life 2021-04 and Node 15 end-of-life 2021-06-01)
* Switch to [pure ES modules](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c#how-can-i-move-my-commonjs-project-to-esm)? I haven't investigated how/if this would work yet
* Update rule options:
  * Update [consistent-output](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/consistent-output.md) rule option default to `always` since it's always best practice to assert the `output`:
    * [As of ESLint 7](https://eslint.org/docs/user-guide/migrating-to-7.0.0#additional-validation-added-to-the-ruletester-class), the `output` property is required for test cases that provide an autofix
    * And even when test cases don't autofix, it's still best practice to assert that there's no autofix with `output: null`
  * Add `meta.hasSuggestions` property to [meta-property-ordering](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/meta-property-ordering.md) rule (https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/pull/119)
* Add new `recommended` rules:
  * I would actually add ALL existing rules. I've used all of this plugin's rules successfully in a number of my eslint plugins, and I find them pretty much all useful and agreeable. Only about 1/3 of existing rules are enabled as recommended right now, and I see this as a missed opportunity, since most people only bother to enable the recommended rules. Enabling more rules means more exposure/battle-hardening of the rules and more benefit to users.

**Comment by bradzacher** [comment]
> I would actually add ALL existing rules. I've used all of this plugin's rules successfully in a number of my eslint plugins, and I find them pretty much all useful and agreeable.

The problem you've got is "pretty much".
The recommended set needs to be unopinionated.
The last thing you want is for someone to disable a rule that's recommended.

Examples of rules that are opinionated: require-meta-type, report-message-format, test-case-property-ordering, test-case-shorthand-strings.

Speaking from experience with @typescript-eslint - you want the recommended set to just be an unopinionated set of rules that catch **problems**.

**Comment by aladdin-add** [comment]
there is already a preset so you can extend like:
```js
extends: ["plugin:eslint-plugin/all"]
```

**Comment by bmish** [comment]
@bradzacher if the goal is to leave "opinionated" or "stylistic" rules (i.e. about the ordering of properties) out of the `recommended` config, I'd be fine with that.

Here are the rules I would specifically advocate for being added to the `recommended` config because they improve rule usability or facilitate eslint plugin development:

| Rule | Why it could make sense in the `recommended` config |
| --- | --- |
| [consistent-output](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/consistent-output.md) | asserting output is required by ESLint 7 when there's an autofix, and even when there's no autofix, it's still valuable to assert that there's no autofix |
| [no-deprecated-context-methods](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/no-deprecated-context-methods.md) | we should help people avoid deprecated APIs |
| [prefer-object-rule](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/prefer-object-rule.md) | the old rule style is deprecated |
| [prefer-output-null](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/prefer-output-null.md) | it's significantly easier to distinguish if a test case has an autofix or not when using `output: null` |
| [prefer-placeholders](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/prefer-placeholders.md) | the rule doc explains benefits |
| [prefer-replace-text](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/prefer-replace-text.md) | improved readability |
| [require-meta-docs-description](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/require-meta-docs-description.md) | it's easier for users to understand rules when they have descriptions |
| [require-meta-docs-url](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/require-meta-docs-url.md) | it's easier for users to understand rules when they provide a link to their documentation |
| [require-meta-has-suggestions](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/require-meta-has-suggestions.md) | this new property will be required for suggestable rules in ESLint 8 |
| [require-meta-schema](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/require-meta-schema.md) | enforcing rule schemas reduces the risk of users configuring a rule incorrectly, and even rules that have no options should provide `schema: []` so that eslint can validate that no options are passed to them |
| [require-meta-type](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/require-meta-type.md) | this enables users to choose what type of rule they want to run (based on the three types used by eslint core: `problem`, `suggestion`, or `layout`). However, if users commonly use types other than these, then I agree it would make sense to exclude this rule. |

**Comment by aladdin-add** [comment]
agreed with @bradzacher .

eslint has set a very high bar to `eslint:recommended` (https://github.com/eslint/eslint/issues/14673). I think better to adopt the similar policy in the plugin. I'm 👍 to the following rules:
* consistent-output
* no-deprecated-context-methods
* prefer-object-rule
* require-meta-has-suggestions
* require-meta-type

**Comment by bmish** [comment]
Sounds good. Can I ask why you would exclude these ones:

* `require-meta-docs-url`
* `require-meta-schema`

I consider these two especially high impact in that they both improve external-facing rule usability by helping developers understand lint rules (by providing an IDE documentation link) and ensuring developers configure rules correctly.

**Comment by aladdin-add** [comment]
yes, it is a best practice, and should be enabled in most cases, but there are some exceptions, e.g. internal-rules: https://github.com/eslint/eslint/tree/master/tools/internal-rules

**Comment by bradzacher** [comment]
Example: none of the ~400 internal rules at Facebook have a schema defined, because they take no options. Nor do they have a docs url defined, because we have our own infra for dealing with rule docs.

**Comment by bmish** [comment]
Having no rule options is totally fine. The [require-meta-schema](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/require-meta-schema.md) rule doc explains that `schema: []` should be specified for rules that do not have options in order to enforce that no options are accidentally passed to such rules, and rules that specify `schema: []` are allowed by this rule.

**Comment by bmish** [comment]
Thanks to both of you for bringing up some use cases where a plugin might not want to specify rule URLs. Those are helpful.

**Comment by bmish** [comment]
I have an update to share regarding the [`require-meta-schema`](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/require-meta-schema.md) rule. I have recommended to the ESLint team that [schemas should be required for rules with options](https://github.com/eslint/eslint/issues/14709) and they are interested in this idea so I'll be opening an RFC for it targeting ESLint 9.

With this change, rules without options will no longer need to specify `schema: []`, as that will be the default behavior. That means we can update the `require-meta-schema` rule to only require the `schema` property for rules that actually have options (which is the same as how [`require-meta-fixable`](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/require-meta-fixable.md) only requires the `fixable` property for rules with fixers).

As a result of this change, we should be able to enable the `require-meta-schema` rule as `recommended` in eslint-plugin-eslint-plugin v4.

**Comment by aladdin-add** [comment]
As eslint v8.0.0 will be released in a few weeks, we can start merging these PRs, and made a prerelease version. thoughts? @bmish

**Comment by bmish** [comment]
@aladdin-add I believe all PRs for the next release are ready, except for #179 which I can try to get to soon. A prerelease version is a good idea.

**Comment by bmish** [comment]
Two more notes:
* We should release a minor version first with these two non-breaking changes: #178 and #179 (once implemented).
* We should also consider including TypeScript support (#176) in this release. TS support will build on top of ESM rule support (#177) and I could give it a try, but #177 needs to be merged first for me to do so.

**Comment by bmish** [comment]
Open question: I have added a `requireSchemaPropertyWhenOptionless` option (default true to maintain current behavior) to [require-meta-schema](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/require-meta-schema.md) to address the above concerns. We have a few paths going forward:
1. Leave option enabled for v4 release. Disable it once ESLint changes their default behavior to require schemas for rules with options (since then it will no longer be necessary to specify `schema: []` for rules without options).
2. Disable the option in v4 release.

**Comment by aladdin-add** [comment]
re requireSchemaPropertyWhenOptionless: I'm fine either way. which one do you prefer?

**Comment by bmish** [comment]
I personally think we should enforce the more strict behavior (keep `requireSchemaPropertyWhenOptionless` enabled by default) for now until it becomes no longer necessary (as a result of a potential future ESLint change).

**Comment by bmish** [comment]
I realized that a lot of our existing rules need to be updated to handle suggestions. I think it's pretty important that we correct this gap in coverage in this major release, as the suggestions feature has been available for nearly 2 years now. I've opened issues to track these gaps, and I will try to open PRs to address these in the coming days. The changes required should not be that large.

- #190 
- #191 
- #192 
- #193

**Comment by aladdin-add** [comment]
👍 nice catch!

**Comment by aladdin-add** [comment]
@bmish I've tested it in a few plugins - all working fine, an excellent job! 🎉

just pushed 4.0.0 - let's wait and see if need patch releases. 😄
